docs/source/onedal/algorithms/nearest-neighbors/knn.rst

```diff
@@ -0,0 +1,167 @@
+.. ******************************************************************************
+.. * Copyright 2020-2021 Intel Corporation
+.. *
+.. * Licensed under the Apache License, Version 2.0 (the "License");
+.. * you may not use this file except in compliance with the License.
+.. * You may obtain a copy of the License at
+.. *
+.. *     http://www.apache.org/licenses/LICENSE-2.0
+.. *
+.. * Unless required by applicable law or agreed to in writing, software
+.. * distributed under the License is distributed on an "AS IS" BASIS,
+.. * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+.. * See the License for the specific language governing permissions and
+.. * limitations under the License.
+.. *******************************************************************************/
+
+.. highlight:: cpp
+.. default-domain:: cpp
+
+.. _alg_knn:
+
+===================================================
+k-Nearest Neighbors Classification and Search(k-NN)
+===================================================
+
+.. include:: ../../../includes/nearest-neighbors/knn-introduction.rst
+
+------------------------
+Mathematical formulation
+------------------------
+
+.. _knn_t_math:
+
+Training
+--------
+.. tabs::
+
+  .. group-tab:: Classification
+
+   Let :math:`X = \{ x_1, \ldots, x_n \}` be the training set of
+   :math:`p`-dimensional feature vectors, let :math:`Y = \{ y_1, \ldots, y_n \}` be
+   the set of class labels, where :math:`y_i \in \{ 0, \ldots, c-1 \}`, :math:`1
+   \leq i \leq n`. Given :math:`X`, :math:`Y` and the number of nearest neighbors
+   :math:`k`, the problem is to build a model that allows distance computation
+   between the feature vectors in training and inference sets at the inference
+   stage.
+
+  .. group-tab:: Search
+
+   Let :math:`X = \{ x_1, \ldots, x_n \}` be the training set of
+   :math:`p`-dimensional feature vectors. Given :math:`X` and the number of nearest neighbors
+   :math:`k`, the problem is to build a model that allows distance computation
+   between the feature vectors in training and inference sets at the inference
+   stage.
+
+.. _knn_t_math_brute_force:
+
+Training method: *brute-force*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The training operation produces the model that stores all the feature vectors
+from the initial training set :math:`X`.
+
+.. _knn_t_math_kd_tree:
+
+Training method: *k-d tree*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The training operation builds a :math:`k`-:math:`d` tree that partitions the
+training set :math:`X` (for more details, see :txtref:`k-d Tree <kd_tree>`).
+
+.. _knn_i_math:
+
+Inference
+---------
+.. tabs::
+
+  .. group-tab:: Classification
+
+   Let :math:`X' = \{ x_1', \ldots, x_m' \}` be the inference set of
+   :math:`p`-dimensional feature vectors. Given :math:`X'`, the model produced at
+   the training stage and the number of nearest neighbors :math:`k`, the problem is
+   to predict the label :math:`y_j'` for each :math:`x_j'`, :math:`1 \leq j \leq
```

Comment:
don't you need to mention that y_j is from the Y set?

Reply:
done